File: Two Pointers/18. 4Sum/Solution.py
from typing import List

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        # M : len(result), N : len(nums)
        # Time Complexity: O(N logN + N^4 + M)
        # Space Complexity: O(M)
        # n = len(nums)
        # nums.sort()
        # result = set()
        # for a in range(n):
        #     for b in range(a + 1, n):
        #         for c in range(b + 1, n):
        #             for d in range(c + 1, n):
        #                 if nums[a] + nums[b] + nums[c] + nums[d] == target:
        #                     result.add((
        #                         nums[a], 
        #                         nums[b], 
        #                         nums[c], 
        #                         nums[d]
        #                     ))
        # return list(result)


        # M : len(result), N : len(nums)
        # Time Complexity: O(N logN + N^3)
        # Space Complexity: O(M)
        n = len(nums)
        nums.sort()
        result = []
        prevI = float('inf')
        for i in range(n):
            if prevI == nums[i]:
                prevI = nums[i]
                continue
            prevI = nums[i]
            prevJ = float('inf')

            for j in range(i + 1, n):
                if prevJ == nums[j]:
                    prevJ = nums[j]
                    continue
                prevJ = nums[j]

                l, r = j + 1, n - 1
                while l < r:
                    summ = nums[i] + nums[j] + nums[l] + nums[r]
                    if summ < target:
                        l += 1
                    elif summ > target:
                        r -= 1
                    else:
                        result.append([nums[i], nums[j], nums[l], nums[r]])
                        l += 1
                        r -= 1
                        while l < r and nums[l] == nums[l - 1]:
                            l += 1
        return result

File: Two Pointers/18. 4Sum/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        // M : len(result), N : len(nums)
        // Time Complexity: O(N logN + N^4 + M)
        // Space Complexity: O(M)
        // int n = nums.length;
        // Arrays.sort(nums);
        // Set<List<Integer>> result = new HashSet<>();
        // for (int a=0; a<n; a++){
        //     for (int b=a+1; b<n; b++){
        //         for (int c=b+1; c<n; c++){
        //             for (int d=c+1; d<n; d++){
        //                 int summ = nums[a] + nums[b] + nums[c] + nums[d];
        //                 if (summ != target) continue;
        //                 List<Integer> temp = Arrays.asList(nums[a], nums[b], nums[c], nums[d]);
        //                 result.add(temp);
        //             }
        //         }
        //     }
        // }
        // return new ArrayList<>(result);


        // M : len(result), N : len(nums)
        // Time Complexity: O(N logN + N^3)
        // Space Complexity: O(M)
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int prevI = Integer.MIN_VALUE;
        for (int i=0; i<n; i++){
            if (prevI == nums[i]){
                prevI = nums[i];
                continue;
            }
            prevI = nums[i];
            int prevJ = Integer.MIN_VALUE;
            for (int j=i+1; j<n; j++){
                if (prevJ == nums[j]){
                    prevJ = nums[j];
                    continue;
                }
                prevJ = nums[j];
                int l = j + 1;
                int r = n - 1;
                while (l < r){
                    long summ = (long)nums[i] + (long)nums[j] + (long)nums[l] + (long)nums[r];
                    if (summ < target){
                        l++;
                    } else if (summ > target){
                        r--;
                    } else{
                        result.add(Arrays.asList(nums[i], nums[j], nums[l], nums[r]));
                        l++;
                        r--;
                        while (l < r && nums[l] == nums[l - 1]) {
                            l++;
                        }
                    }
                }
            }
        }
        return result;
    }
}